README.md
Oaf
---

Care-free web app prototyping using files and scripts

[![Gem Version](https://badge.fury.io/rb/oaf.png)](http://badge.fury.io/rb/oaf)
[![Build Status](https://travis-ci.org/ryanuber/oaf.png)](https://travis-ci.org/ryanuber/oaf)
[![Coverage Status](https://coveralls.io/repos/ryanuber/oaf/badge.png)](https://coveralls.io/r/ryanuber/oaf)
[![Code Climate](https://codeclimate.com/github/ryanuber/oaf.png)](https://codeclimate.com/github/ryanuber/oaf)

[Documentation](http://rubydoc.info/gems/oaf/frames)

`Oaf` provides stupid-easy way of creating dynamic web applications by setting
all best practices and security considerations aside until you are sure that you
want to invest your time doing so.

`Oaf` was created as a weekend project to create a small, simple HTTP server
program that uses script execution as its primary mechanism for generating
responses.

Example
-------

Create an executable file named `hello.GET`:

```
#!/bin/bash
echo "Hello, ${USER}!"
```

Start the server by running the `oaf` command, then make a request:

```
$ curl localhost:9000/hello
Hello, ryanuber!
```

Installation
------------

Oaf is available on [rubygems](http://rubygems.org/gems/oaf), which means you
can do:

```
gem install oaf
```

### Accepted files

`Oaf` will run *ANY* file with the executable bit set, be it shell, Python, Ruby,
compiled binary, or whatever else you might have.

`Oaf` can also use plain text files.

### How file permissions affect output

* If the file in your request is executable, the output of its execution will
  be used as the return data.
* If the file is *NOT* executable, then the contents of the file will be used
  as the return data.

### Nested methods

You can create nested methods using simple directories. Example:

```
$ ls ./examples/
hello.GET

$ curl http://localhost:8000/examples/hello
Hello, world!
```

### HTTP Methods
Files must carry the extension of the HTTP method used to invoke them.
Oaf should support any HTTP method, including custom methods.

### Headers and Status
You can indicate HTTP headers and status using stdout from your script.

```
#!/bin/bash
cat <<EOF
Hello, world!
---
content-type: text/plain
200
EOF
```

Separated by 3 dashes on a line of their own (`---`), the very last block
of output can contain headers and response status.

### Getting request headers, query parameters, and body

Headers, query parameters, and request body are all passed to executables using
the environment. To defeat overlap in variables, they are namespaced using a
prefix. The environment variables are as follows:

* Headers: `oaf_header_*`
* Request parameters: `oaf_param_*`
* Request body: `oaf_request_body`
* Request path: `oaf_request_path`

Below is a quick example of a shell script which makes use of the request data:

```
#!/bin/bash
echo "You passed the Accept header: $oaf_header_accept"
echo "You passed the 'message' value: $oaf_param_message"
echo "You passed the request body: $oaf_request_body"
echo "This method is located at: $oaf_request_path"
```

Headers query parameter names are converted to all-lowercase, and dashes are
replaced with underscores. This is due to the way the environment works. For
example, if you wanted to get at the `Content-Type` header, you could with the
environment variable `$oaf_header_content_type`.

### Catch-all methods

Catch-all's can be defined by naming a file inside of a directory, beginning and
ending with underscores (`_`). So for example, `test/_default_.GET` will match:
`GET /test/anything`, `GET /test/blah`, etc.

If you want to define a top-level method for `/test`, you would do so in the
file at `/test.GET`.

Q&A
---

**Why are the headers and status at the bottom of the response?**
Because it is much easier to echo these last. Since Oaf reads response
data directly from stdout/stderr, it is very easy for debug or error messages
to interfere with them. By specifying the headers and status last, we minimize
the chances of unexpected output damaging the response, as all processing is
already complete.

**Why the name `Oaf`?**
It's a bit of a clumsy and "oafish" approach at web application prototyping. I
constantly find myself trying to write server parts of programs before I have
even completed basic functionality, and sometimes even before I have a clear
idea of what it is I want the program to do.

Acknowledgements
----------------

`Oaf` is inspired by [Stubb](https://github.com/knuton/stubb). A number of ideas
and conventions were borrowed from it. Kudos to
[@knuton](https://github.com/knuton) for having a great idea.


debian/changelog
oaf (0.3.1) UNRELEASED; urgency=low

  * Initial release.

 -- Ryan Uber <ru@ryanuber.com>  Thu, 09 Jan 2014 14:32:22 -0800


spec/http-spec.rb
# oaf - Care-free web app prototyping using files and scripts
# Copyright 2013 Ryan Uber <ru@ryanuber.com>
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

require 'oaf'
require 'ostruct'

module Oaf
  describe "Returning HTTP Responses" do
    it "should return safe defaults if output is empty" do
      headers, status, body = Oaf::HTTP.parse_response ''
      headers.should eq({})
      status.should eq(200)
      body.should eq("\n")
    end

    it "should return safe defaults when only body is present" do
      text = "This is a test\n"
      headers, status, body = Oaf::HTTP.parse_response text
      headers.should eq({})
      status.should eq(200)
      body.should eq("This is a test\n")
    end

    it "should return headers correctly" do
      text = "---\nx-powered-by: oaf"
      headers, status, body = Oaf::HTTP.parse_response text
      headers.should eq({'x-powered-by' => 'oaf'})
    end

    it "should return status correctly" do
      text = "---\n201"
      headers, status, body = Oaf::HTTP.parse_response text
      status.should eq(201)
    end

    it "should return body correctly" do
      text = "This is a test\n---\n200"
      headers, status, body = Oaf::HTTP.parse_response text
      body.should eq("This is a test\n")
    end

    it "should return body correctly when no metadata is present" do
      text = "This is a test"
      headers, status, body = Oaf::HTTP.parse_response text
      body.should eq("This is a test\n")
    end
  end

  describe "Running an HTTP Server" do
    before(:all) do
      require 'webrick-mocks'
      @tempdir1 = Dir.mktmpdir
      @f1 = Tempfile.new ['oaf', '.GET'], @tempdir1
      @f1.write "This is a test.\n---\n201\nx-powered-by: oaf"
      @f1.close
      @f1request = File.basename(@f1.path).sub!(/\.GET$/, '')

      @f2 = Tempfile.new ['oaf', '.PUT'], @tempdir1
      @f2.write "Containable Test\n---\n202\nx-powered-by: oaf"
      @f2.close
      @f2request = File.basename(@f2.path).sub!(/\.PUT$/, '')
    end

    after(:all) do
      @f1.delete
      @f2.delete
      Dir.delete @tempdir1
    end

    before(:each) do
      @webrick = double()
      @webrick.should_receive(:start).once.and_return(true)
      WEBrick::HTTPServer.stub(:new).and_return(@webrick)
    end

    it "should start an HTTP server" do
      @webrick.should_receive(:mount_proc).with('/').once \
        .and_yield(Oaf::FakeReq.new, Oaf::FakeRes.new)
      Oaf::HTTP.serve '/tmp', 9000
    end

    it "should parse the request properly" do
      req = Oaf::FakeReq.new :path => @f1request
      res = Oaf::FakeRes.new
      @webrick.should_receive(:mount_proc).with('/').once \
        .and_yield(req, res)
      Oaf::HTTP.serve @tempdir1, 9000
      res.body.should eq("This is a test.\n")
      res.status.should eq(201)
      res.header.should eq('x-powered-by' => 'oaf')
    end

    it "should accept containable methods properly" do
      req = Oaf::FakeReq.new({:path => @f2request, :method => 'PUT'})
      res = Oaf::FakeRes.new
      @webrick.should_receive(:mount_proc).with('/').once \
        .and_yield(req, res)
      Oaf::HTTP.serve(@tempdir1, 9000)
      res.body.should eq("Containable Test\n")
      res.status.should eq(202)
      res.header.should eq('x-powered-by' => 'oaf')
    end
  end
end


spec/util-spec.rb
# oaf - Care-free web app prototyping using files and scripts
# Copyright 2013 Ryan Uber <ru@ryanuber.com>
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

require 'oaf'
require 'ostruct'
require 'tempfile'

module Oaf
  describe "Header Lines" do
    it "should detect a valid header line" do
      result = Oaf::Util.is_http_header? 'x-powered-by: oaf'
      result.should be_true
    end

    it "should detect an invalid header line" do
      result = Oaf::Util.is_http_header? 'invalid header line'
      result.should be_false
    end

    it "should parse a name and value from a header line" do
      result = Oaf::Util.get_http_header 'x-powered-by: oaf'
      result.should eq(['x-powered-by', 'oaf'])
    end

    it  "should detect an invalid header line during header parsing" do
      result = Oaf::Util.get_http_header 'invalid header line'
      result.should be_nil
    end
  end

  describe "Status Lines" do
    it "should detect all valid HTTP status codes" do
      [200, 201, 202, 203, 204, 205, 206,
       300, 301, 302, 303, 304, 305, 306, 307,
       400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411,
       412, 413, 414, 415, 416, 417,
       500, 501, 502, 503, 504, 505].each do  |status|
        result = Oaf::Util.is_http_status? status
        result.should be_true
      end
    end

    it "should be able to validate a string http status line" do
      result = Oaf::Util.is_http_status? '200'
      result.should be_true
    end

    it "should detect an invalid http status code" do
      result1 = Oaf::Util.is_http_status? 199
      result2 = Oaf::Util.is_http_status? 'not a status'
      result1.should be_false
      result2.should be_false
    end

    it "should retrieve an http status code" do
      result = Oaf::Util.get_http_status '200'
      result.should eq(200)
    end

    it "should detect an invalid status code during parsing" do
      result = Oaf::Util.get_http_status 'not a status'
      result.should be_nil
    end
  end

  describe "Format Request Headers" do
    it "should return a single key/value for just one header" do
      headers = [['x-powered-by', 'oaf']]
      result = Oaf::Util.format_request_headers headers
      result.should eq('x-powered-by:oaf')
    end

    it "should return a comma-delimited list for multiple headers" do
      headers = [['x-powered-by', 'oaf'], ['content-type', 'text/plain']]
      result = Oaf::Util.format_request_headers headers
      result.should eq('x-powered-by:oaf,content-type:text/plain')
    end

    it "should return nil if no headers present" do
      headers = []
      result = Oaf::Util.format_request_headers headers
      result.should be_nil
    end
  end

  describe "Parse Request Metadata From Output" do
    it "should find headers in request metadata" do
      text = ['---', 'x-powered-by: oaf', '201'].join "\n"
      headers, status, size = Oaf::Util.parse_http_meta text
      headers.should eq('x-powered-by' => 'oaf')
    end

    it "should return the number of lines the metadata consumes" do
      text = ['---', 'x-powered-by: oaf', '200'].join "\n"
      headers, status, size = Oaf::Util.parse_http_meta text
      size.should eq(3)
    end

    it "should assume 200 as the default return code" do
      text = ['---', 'x-powered-by: oaf'].join "\n"
      headers, status, size = Oaf::Util.parse_http_meta text
      status.should eq(200)
    end

    it "should assume meta size 0 if no metadata is present" do
      text = 'this response uses default metadata'
      headers, status, size = Oaf::Util.parse_http_meta text
      size.should eq(0)
    end

    it "should return safe defaults if the response is empty" do
      text = ''
      headers, status, size = Oaf::Util.parse_http_meta text
      headers.should eq({})
      status.should eq(200)
      size.should eq(0)
    end
  end

  describe "Determine File Paths" do
    before(:all) do
      @tempdir1 = Dir.mktmpdir
      @tempdir2 = Dir.mktmpdir

      @f1 = Tempfile.new ['oaf', '.GET'], @tempdir1
      @f1.write "This is a test.\n"
      @f1.close
      @f1request = File.basename(@f1.path).sub!(/\.GET$/, '')

      @f2 = Tempfile.new ['_', '_.GET'], @tempdir1
      @f2.write "This is a default file.\n"
      @f2.close
    end

    after(:all) do
      @f1.delete
      @f2.delete
      Dir.delete @tempdir1
      Dir.delete @tempdir2
    end

    it "should find existing files correctly" do
      result = Oaf::Util.get_request_file @tempdir1, @f1request, 'GET'
      result.should eq(@f1.path)
    end

    it "should return the fall-through file if request file doesn't exist" do
      result = Oaf::Util.get_request_file @tempdir1, 'nonexistent', 'GET'
      result.should eq(@f2.path)
    end

    it "should return nil if neither the requested or default file exist" do
      result = Oaf::Util.get_request_file @tempdir2, 'nonexistent', 'GET'
      result.should be_nil
    end
  end

  describe "Executing and Reading Files" do
    before(:all) do
      @f1 = Tempfile.new 'oaf'
      @f1.chmod 0755
      @f1.write "#!/bin/bash\necho 'This is a test'"
      @f1.close

      @f2 = Tempfile.new 'oaf'
      @f2.chmod 0644
      @f2.write "This is a test\n"
      @f2.close
    end

    after(:all) do
      @f1.delete
      @f2.delete
    end

    it "should execute a file if it is executable" do
      result = Oaf::Util.get_output @f1.path, nil, nil
      result.should eq("This is a test\n")
    end

    it "should read file contents if it is not executable" do
      result = Oaf::Util.get_output @f2.path, nil, nil
      result.should eq("This is a test\n")
    end

    it "should assume safe defaults if the file doesnt exist" do
      result = Oaf::Util.get_output nil, nil, nil
      result.should eq(Oaf::Util.get_default_response)
    end
  end
end


spec/spec_helper.rb
require 'simplecov'
require 'coveralls'
Coveralls.wear!
SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
  SimpleCov::Formatter::HTMLFormatter,
  Coveralls::SimpleCov::Formatter
]
SimpleCov.start do
  add_filter '/spec/'
end
require 'oaf'
require 'ostruct'
require 'tempfile'


lib/oaf.rb
# oaf - Care-free web app prototyping using files and scripts
# Copyright 2013 Ryan Uber <ru@ryanuber.com>
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

require 'oaf/version'
require 'oaf/util'
require 'oaf/httphandler'
require 'oaf/httpserver'


lib/oaf/http.rb
# oaf - Care-free web app prototyping using files and scripts
# Copyright 2013 Ryan Uber <ru@ryanuber.com>
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

require 'webrick'

module Oaf

  module HTTP
    extend Oaf
    extend self

    # Given output from a script, parse HTTP response details and return them
    # as a hash, including body, status, and headers.
    #
    # == Parameters:
    # output::
    #   The output text data returned by a script.
    #
    # == Returns:
    # A hash containing the HTTP response details (body, status, and headers).
    #
    def parse_response output
      has_meta = false
      headers = {'content-type' => 'text/plain'}
      status = 200
      headers, status, meta_size = Oaf::Util.parse_http_meta output
      lines = output.split("\n")
      body = lines.take(lines.length - meta_size).join("\n")+"\n"
      [headers, status, body]
    end

    # Safely retrieves the request body, and assumes an empty string if it
    # cannot be retrieved. This helps get around a nasty exception in WEBrick.
    #
    # == Parameters:
    # req::
    #   A WEBrick::HTTPRequest object
    #
    # == Returns:
    # A string containing the request body
    #
    def get_request_body req
      if ['POST', 'PUT'].member? req.request_method
        begin
          result = req.body
        rescue WEBrick::HTTPStatus::LengthRequired
          result = ''  # needs to be in rescue for coverage
        end
      end
      result
    end

    # Consume HTTP response details and set them into a response object.
    #
    # == Parameters:
    # res::
    #   A WEBrick::HTTPResponse object
    # headers::
    #   A hash containing HTTP response headers
    # body::
    #   A string containing the HTTP response body
    # status::
    #   An integer indicating the response status
    #
    def set_response! res, headers, body, status
      headers.each do |name, value|
        res.header[name] = value
      end
      res.body = body
      res.status = status
    end

    # Invokes the Webrick web server library to handle incoming requests, and
    # routes them to the appropriate scripts if they exist on the filesystem.
    #
    # == Parameters:
    # path::
    #   The path in which to search for files
    # port::
    #   The TCP port to listen on
    #
    def serve path, port
      server = WEBrick::HTTPServer.new :Port => port
      server.mount_proc '/' do |req, res|
        req_headers = req.header
        req_query = req.query
        req_body = Oaf::HTTP::get_request_body req
        file = Oaf::Util.get_request_file path, req.path, req.request_method
        out = Oaf::Util.get_output file, req_headers, req_body, req_query
        res_headers, res_status, res_body = Oaf::HTTP.parse_response out
        Oaf::HTTP.set_response! res, res_headers, res_body, res_status
      end
      trap 'INT' do server.shutdown end
      server.start
    end
  end
end


lib/oaf/version.rb
# oaf - Care-free web app prototyping using files and scripts
# Copyright 2013 Ryan Uber <ru@ryanuber.com>
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

module Oaf
  VERSION = '0.3.1'
end


lib/oaf/httpserver.rb
# oaf - Care-free web app prototyping using files and scripts
# Copyright 2013 Ryan Uber <ru@ryanuber.com>
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

require 'oaf/util'
require 'oaf/httphandler'
require 'webrick'

module Oaf::HTTPServer
  extend self

  # Given output from a script, parse HTTP response details and return them
  # as a hash, including body, status, and headers.
  #
  # == Parameters:
  # output::
  #   The output text data returned by a script.
  #
  # == Returns:
  # A hash containing the HTTP response details (body, status, and headers).
  #
  def parse_response output
    has_meta = false
    headers = {'content-type' => 'text/plain'}
    status = 200
    headers, status, meta_size = Oaf::Util.parse_http_meta output
    lines = output.split("\n")
    body = lines.take(lines.length - meta_size).join("\n")+"\n"
    [headers, status, body]
  end

  # Safely retrieves the request body, and assumes an empty string if it
  # cannot be retrieved. This helps get around a nasty exception in WEBrick.
  #
  # == Parameters:
  # req::
  #   A WEBrick::HTTPRequest object
  #
  # == Returns:
  # A string containing the request body
  #
  def get_request_body req
    if ['POST', 'PUT'].member? req.request_method
      begin
        result = req.body
      rescue WEBrick::HTTPStatus::LengthRequired
        result = ''  # needs to be in rescue for coverage
      end
    end
    result
  end

  # Consume HTTP response details and set them into a response object.
  #
  # == Parameters:
  # res::
  #   A WEBrick::HTTPResponse object
  # headers::
  #   A hash containing HTTP response headers
  # body::
  #   A string containing the HTTP response body
  # status::
  #   An integer indicating the response status
  #
  def set_response! res, headers, body, status
    headers.each do |name, value|
      res.header[name] = value
    end
    res.body = body
    res.status = status
  end

  # Invokes the Webrick web server library to handle incoming requests, and
  # routes them to the appropriate scripts if they exist on the filesystem.
  #
  # == Parameters:
  # path::
  #   The path in which to search for files
  # port::
  #   The TCP port to listen on
  #
  def serve path, port
    server = WEBrick::HTTPServer.new :Port => port
    server.mount '/', Oaf::HTTPHandler, path
    trap 'INT' do server.shutdown end
    server.start
  end
end


lib/oaf/util.rb
# oaf - Care-free web app prototyping using files and scripts
# Copyright 2013 Ryan Uber <ru@ryanuber.com>
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

module Oaf::Util
  extend self

  # Determines if a line of output looks anything like an HTTP header
  # declaration.
  #
  # == Parameters:
  # line::
  #   A line of text to examine
  #
  # == Returns:
  # A boolean, true if it can be read as a header string, else false.
  #
  def is_http_header? line
    line.split(':').length == 2
  end

  # Retrieves a hash in the form `name` => `value` from a string that
  # describes an HTTP response header.
  #
  # == Parameters:
  # line::
  #   A line of text to parse
  #
  # == Returns
  # A hash in the form `name` => `value`, or `nil`
  #
  def get_http_header line
    return nil if not is_http_header? line
    parts = line.split(':')
    {parts[0].strip => parts[1].strip}
  end

  # Retrieves the numeric value from a line of text as an HTTP status code.
  #
  # == Parameters:
  # line::
  #   A line of text to parse
  #
  # == Returns:
  # An integer if valid, else `nil`.
  #
  def get_http_status line
    is_http_status?(line) ? line.to_i : nil
  end

  # Determines if an HTTP status code is valid per RFC2616
  #
  # == Parameters:
  # code::
  #   A number to validate
  #
  # == Returns
  # A boolean, true if valid, else false.
  #
  def is_http_status? code
    (200..206).to_a.concat((300..307).to_a).concat((400..417).to_a) \
      .concat((500..505).to_a).include? code.to_i
  end

  # Convert various pieces of data about the request into a single hash, which
  # can then be passed on as environment data to a script.
  #
  # == Parameters:
  # req_path::
  #   A string with the HTTP request path
  # headers::
  #   A hash of request headers
  # params::
  #   A hash of request parameters
  # body::
  #   A string containing the request body
  #
  # == Returns:
  # A flat hash containing namespaced environment parameters
  #
  def prepare_environment req_path, headers, params, body
    result = Hash.new
    {'header' => headers, 'param' => params}.each do |prefix, data|
      data.each do |name, value|
        result.merge! Oaf::Util.environment_item prefix, name, value
      end
    end
    result.merge! Oaf::Util.environment_item 'request', 'path', req_path
    result.merge Oaf::Util.environment_item 'request', 'body', body
  end

  # Prepares a key for placement in the execution environment. This includes
  # namespacing variables and converting characters to predictable and
  # easy-to-use names.
  #
  # == Parameters:
  # prefix::
  #   A prefix for the key. This helps with separation.
  # key::
  #   The key to sanitize
  #
  # == Returns:
  # A string with the prepared value
  #
  def prepare_key prefix, key
    "oaf_#{prefix}_#{key.gsub('-', '_').downcase}"
  end

  # Formats a single environment item into a hash, which can be merged into a
  # collective environment mapping later on.
  #
  # == Parameters:
  # prefix::
  #   The prefix for the type of item being added.
  # key::
  #   The key name of the environment property
  # value::
  #   The value for the environment property
  #
  # == Returns:
  # A hash with prepared values ready to merge into an environment hash
  #
  def environment_item prefix, key, value
    {Oaf::Util.prepare_key(prefix, key) => Oaf::Util.flatten(value)}
  end

  # Flatten a hash or array into a string. This is useful for preparing some
  # data for passing in via the environment, because multi-dimension data
  # structures are not supported for that.
  #
  # == Parameters:
  # data::
  #   The data to flatten
  #
  # == Returns:
  # A flattened string. It will be empty if the object passed in was not
  # flatten-able.
  #
  def flatten data
    result = ''
    if data.kind_of? Hash
      data.each do |key, val|
        val = Oaf::Util.flatten val if not val.kind_of? String
        result += "#{key}#{val}"
      end
    elsif data.kind_of? Array
      data.each do |item|
        item = Oaf::Util.flatten item if not item.kind_of? String
        result += item
      end
    elsif data.kind_of? String
      result = data
    else
      result = ''
    end
    result
  end

  # Given an array of text lines, iterate over each of them and determine if
  # they may be interpreted as headers or status. If they can, add them to
  # the result.
  #
  # == Parameters:
  # text::
  #   Plain text data to examine
  #
  # == Returns:
  # A 3-item structure containing headers, status, and the number of lines
  # which the complete metadata (including the "---" delimiter) occupies.
  #
  def parse_http_meta text
    headers = {}
    status = 200
    size = 0
    parts = text.split /^---$/
    if parts.length > 1
      for part in parts.last.split "\n"
        if Oaf::Util.is_http_header? part
          headers.merge! Oaf::Util.get_http_header part
        elsif Oaf::Util.is_http_status? part
          status = Oaf::Util.get_http_status part
        else next
        end
        size += size == 0 ? 2 : 1  # compensate for delimiter
      end
    end
    [headers, status, size]
  end

  # Return a default response string indicating that nothing could be
  # done and no response files were found.
  #
  # == Returns:
  # A string with response output for parsing.
  #
  def get_default_response
    "oaf: Not Found\n---\n404"
  end

  # Returns the path to the file to use for the request. If the provided
  # file path does not exist, this method will search for a file within
  # the same directory matching the default convention "_*_".
  #
  # == Parameters:
  # root::
  #   The root path to search within.
  # req_path::
  #   The HTTP request path
  # method::
  #   The HTTP method of the request
  #
  # == Returns:
  # The path to a file to use, or `nil` if none is found.
  #
  def get_request_file root, req_path, method
    file = File.join root, "#{req_path}.#{method}"
    if not File.exist? file
      Dir.glob(File.join(File.dirname(file), "_*_.#{method}")).each do |f|
        file = f
        break
      end
    end
    File.exist?(file) ? file : nil
  end

  # Fork a new process, in which we can safely modify the running environment
  # and run a command, sending data back to the parent process using an IO
  # pipe. This can be done in a single line in ruby >= 1.9, but we will do it
  # the hard way to maintain compatibility with older rubies.
  #
  # == Parameters:
  # env::
  #   The environment data to use in the subprocess.
  # command::
  #   The command to execute against the server
  #
  # == Returns:
  # A string of stderr concatenated to stdout.
  #
  def run_buffered env, command
    out, wout = IO.pipe
    pid = fork do
      out.close
      ENV.replace env
      wout.write %x(#{command} 2>&1)
      at_exit { exit! }
    end
    wout.close
    Process.wait pid
    out.read
  end

  # Executes a file, or reads its contents if it is not executable, passing
  # it the request headers and body as arguments, and returns the result.
  #
  # == Parameters:
  # req_path::
  #   The HTTP request path
  # file::
  #   The path to the file to use for output
  # headers::
  #   The HTTP request headers to pass
  # body::
  #   The HTTP request body to pass
  # params::
  #   The HTTP request parameters to pass
  #
  # == Returns:
  # The result from the file, or a default result if the file is not found.
  #
  def get_output file, req_path=nil, headers=[], body=[], params=[]
    if file.nil?
      out = Oaf::Util.get_default_response
    elsif File.executable? file
      env = Oaf::Util.prepare_environment req_path, headers, params, body
      out = Oaf::Util.run_buffered env, file
    else
      out = File.open(file).read
    end
    out
  end
end
